### PR TITLE
feat: add access/delegate capability parser exported from @web3-storage/capabilities

### PR DESCRIPTION
**Pull request by gobengo**
Motivation:
* https://github.com/web3-storage/w3protocol/issues/414

### CODE REVIEW

File: packages/capabilities/src/access.js [hunk 1]
```diff
@@ -108,3 +108,77 @@ export const claim = base.derive({
...
+    nb: {
+      // keys SHOULD be CIDs, but we won't require it in the schema
+      delegations: Schema.dictionary({
+        value: Schema.Link.match(),
+      }),
+    },
+    derives: (claim, proof) => {
+      return (
+        fail(equalWith(claim, proof)) ||
+        // @ts-ignore
```
Comment (gobengo) [inline packages/capabilities/src/access.js:131]: @Gozala I could use some help/pairing trying to not need this

File: packages/capabilities/src/access.js [hunk 2]
```diff
@@ -108,3 +108,86 @@ export const claim = base.derive({
...
+    return fail(equalWith(claim, proof)) || true
+  },
+})
+
+/**
+ * Parsed Capability for access/delegate
+ *
+ * @template {Types.Ability} A
+ * @template {Types.URI} R
+ * @typedef {Types.ParsedCapability<A, R, { delegations?: Types.Failure | Schema.Dictionary<string, Types.Link<unknown, number, number, 0 | 1>> }>} ParsedAccessDelegate
```
Comment (gobengo) [inline packages/capabilities/src/access.js:148]: I use this below on lines 159, 160. @Gozala lmk if there is some better way of inferring this type. It's possible I can't just infer it from `delegate` because it's also used in the definition of `delegate` (derives) so there is a circularity?

Comment (Gozala) [reply packages/capabilities/src/access.js:148]: `delegations` should not be `Failure`,  ucanto will not create parsed capability like that. It will either parse successfully or just fail, it will never partially fail.

Comment (Gozala) [reply packages/capabilities/src/access.js:148]:
```suggestion
 * @typedef {Types.ParsedCapability<A, R, { delegations?: Schema.Dictionary<string, Types.Link<unknown, number, number, 0 | 1>> }>} ParsedAccessDelegate
```

Does this not work ?

Comment (gobengo) [reply packages/capabilities/src/access.js]: @Gozala and I met together, and realized I needed to [add this type assertion](https://github.com/web3-storage/w3protocol/pull/420/commits/4f1f2931cecff1d1d1d29e889c4fdfb63ff3b327#diff-e434cc6c1a699df311a0b2faed199a2ff6b6b291d30f95e20b2ea5abfa7da3d9R125)

Comment (gobengo) [reply packages/capabilities/src/access.js]: I proposed this in ucanto to make the type assertion unnecessary https://github.com/web3-storage/ucanto/pull/215

File: packages/capabilities/src/access.js [hunk 3]
```diff
@@ -108,3 +108,90 @@ export const claim = base.derive({
...
+ * checks that the claimed delegation set is equal to or less than the proven delegation set.
+ * usable with {import('@ucanto/interface').Derives}.
+ *
+ * @param {ParsedAccessDelegate} claim
+ * @param {ParsedAccessDelegate} proof
+ */
+function subsetsNbDelegations(claim, proof) {
+  /** @param {ParsedAccessDelegate} ucan */
+  const nbDelegationsCids = (ucan) =>
+    new Set(Object.values(ucan.nb.delegations || {}).map(String))
```
Comment (Gozala) [inline packages/capabilities/src/access.js:171]: Nit: can you hoist this function, to the top level since it does not enclose anything

File: packages/capabilities/src/access.js [hunk 4]
```diff
@@ -108,3 +108,90 @@ export const claim = base.derive({
...
+  const missingProofs = setDifference(
+    new Set(nbDelegationsCids(claim)),
```
Comment (Gozala) [inline packages/capabilities/src/access.js:173]: You create a set and then pass it into a set constructor, which seems unnecessary

File: packages/capabilities/test/capabilities/access.test.js [hunk 5]
```diff
@@ -371,3 +371,256 @@ describe('access capabilities', function () {
...
+describe('access/delegate', () => {
+  it('can create valid delegations and authorize them', async () => {
+    const issuer = alice
+    const audience = service.withDID('did:web:web3.storage')
+    const bobCanStoreAllWithAlice = await delegate({
+      issuer: alice,
+      audience: bob,
+      capabilities: [{ can: 'store/*', with: alice.did() }],
+    })
+    const validInvocations = [
```
Comment (Gozala) [inline packages/capabilities/test/capabilities/access.test.js:384]: This is really hard to follow and if one of the cases fails the whole test will fail as well. Could you generate tests instead meaning have loop that calls `it` ? It would be easier to follow and failure in one case will not fail the whole suite.

File: packages/capabilities/test/capabilities/access.test.js [hunk 6]
```diff
@@ -371,3 +371,256 @@ describe('access capabilities', function () {
...
+          },
+        })
+        .delegate(),
+      // example from spec, but once each with several different, but all valid, property names to use in the `nb.delegations` dict
+      // https://github.com/web3-storage/specs/blob/7e662a2d9ada4e3fc22a7a68f84871bff0a5380c/w3-access.md?plain=1#L58
+      .../** @type {const} */ ([
+        // correct cid
+        [bobCanStoreAllWithAlice.cid.toString(), bobCanStoreAllWithAlice.cid],
+        // not a cid at all
+        ['thisIsNotACid', bobCanStoreAllWithAlice.cid],
```
Comment (Gozala) [inline packages/capabilities/test/capabilities/access.test.js:402]: 🤔 I'm starting to wonder if the whole dict is a good idea here, but then if we chose list it also may not be ordered.

**Review by Gozala — CHANGES_REQUESTED** [review]
Overall looks good, put bunch of comments but ones I'd like to see addressed before landing are:

1. Remove one layer of set wrapper `new Set(new Set(...`
2. Adjustment to how tests are generated so it's easier to follow what going on.

Thanks

File: packages/capabilities/src/access.js (hunk 4 above)
Comment (gobengo) [reply packages/capabilities/src/access.js]: https://github.com/web3-storage/w3protocol/pull/420/commits/d03e081ef3bfda4e2aeaa01e09aded00b27c5efa

File: packages/capabilities/test/capabilities/access.test.js (hunk 5 above)
Comment (gobengo) [reply packages/capabilities/test/capabilities/access.test.js]: https://github.com/web3-storage/w3protocol/pull/420/commits/edeee6b9d3d619c8468b93c87d1276255effc68f

File: packages/capabilities/src/access.js (hunk 3 above)
Comment (gobengo) [reply packages/capabilities/src/access.js]: hoisted in https://github.com/web3-storage/w3protocol/commit/d03e081ef3bfda4e2aeaa01e09aded00b27c5efa

**Review by Gozala — APPROVED** [review]
Suggested minor changes, feel free to land as is or change a bit and then land.

File: packages/capabilities/test/capabilities/access.test.js [hunk 7]
```diff
@@ -371,3 +371,290 @@ describe('access capabilities', function () {
...
+function createSelfIssuedDelegateInvocation(alice, service) {
+  return Access.delegate
+    .invoke({
+      issuer: alice,
+      audience: service,
+      with: alice.did(),
+      nb: {
+        delegations: {},
+      },
+    })
```
[the remark this answers is not in the document]
Comment (gobengo) [reply packages/capabilities/test/capabilities/access.test.js]: > This function is used just once and does not really improve readability of test.

I removed it https://github.com/web3-storage/w3protocol/pull/420/commits/57f51ea034b7134d7593481445a4f2b8650f3d31

> I would suggest changing it as follows and reusing it all the other tests

I'd prefer not to wrap use of `Access.delegate.invoke()` by other tests unnecessarily, because I find it easier to reason about the other tests if they just use the capability parser directly vs using it through a wrapper like `delegateAccess`
